Solutions/Corvus.Azure.Storage.Tenancy/Corvus/Azure/Storage/Tenancy/ITenantCloudTableFactory.cs

```diff
@@ -0,0 +1,66 @@
+﻿// <copyright file="ITenantCloudTableFactory.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Corvus.Azure.Storage.Tenancy
+{
+    using System.Threading.Tasks;
+    using Corvus.Tenancy;
+    using Microsoft.Azure.Cosmos.Table;
+    using Microsoft.Azure.Storage.Blob;
+
+    /// <summary>
+    /// A factory for a <see cref="CloudTable"/>.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// You use this type to get an instance of an <see cref="CloudTable"/> for a specific
+    /// <see cref="ITenant"/>. It uses a KeyVault to get the storage account key for the tenant, and the
+    /// configuration comes from the tenant via <see cref="TableStorageTenantExtensions.AddTableStorageConfiguration(System.Collections.Generic.IEnumerable{System.Collections.Generic.KeyValuePair{string, object}}, TableStorageTableDefinition, TableStorageConfiguration)"/>.
+    /// </para>
+    /// <para>
+    /// To configure a simple single-tenanted solution, which can ultimately be extended to multitenancy, the easiest route is to configure a configuration-based account key
+    /// provider and configuration for your repositories.
+    /// </para>
+    /// <para>
+    /// First, add the table factory and the configuration account key provider in your table configuration (assuming you have added a standard ConfigurationRoot to your solution).
+    /// </para>
+    /// <code>
+    /// serviceCollection.AddTenantCloudTableFactory();
+    /// serviceCollection.AddTenantConfigurationAccountKeyProvider();
+    /// </code>
+    /// <para>
+    /// <code>
+    /// TenantCloudTableFactory factory;
+    ///
+    /// var repository = await factory.GetTableForTenantAsync(tenantProvider.Root, new TableStorageContainerDefinition("sometable"));
+    /// </code>
+    /// </para>
+    /// <para>
+    /// If you create tables in this way (rather than just newing them up) then your application can easily be multitented
+    /// by ensuring that you always pass the Tenant through your stack, and just default to tenantProvider.Root at the top level.
+    /// </para>
+    /// <para>
+    /// Note that it will be possible for code that obtains a CloudTable in this way to use the resulting object to access
+    /// the CloudTableClient and thus access other tables contained in the same account. As such these objects should only ever be
+    /// handed to trusted code.
+    /// </para>
+    /// <para>
+    /// Note also that because we have not wrapped the resulting CloudTable in a class of our own, we cannot automatically
+    /// implement key rotation.
+    /// </para>
+    /// </remarks>
```

Comment:
Having got to the end here, I find myself uncertain about an important detail: how are tenants separated out? Do you add in the tenant name to the container?

This is an important detail when it comes to table storage, because you have a very limited set of options when it comes to shardability and searchability. If it turns out I'm going to be sharing a container with other tenants I really need to know about that.

I would guess that I don't—I would guess that it two different tenants ask for `foo` they get different actual tables. But I think it'd be good to be clear about that.